packages/vscode-mdx/README.md

```diff
@@ -60,6 +61,113 @@ repository readme.
 Syntax highlighting for MDX is based on the
 [MDX TextMate grammar](https://github.com/wooorm/markdown-tm-language).
 
+### Custom Languages in Code Blocks
+
+MDX for Visual Studio Code supports syntax highlighting for a number of
+well-known languages in code blocks.
+However, it’s impossible to support all languages within the MDX extension.
+Instead, if an extensions adds support for a language, it can add support for
+MDX code blocks.
+
+To support MDX code blocks in your Visual Studio Code extension, use the
+following template.
+Replace `LANGUAGE` with your actual language and remove comments.
+Save the file to `syntaxes/mdx.LANGUAGE.tmLanguage.json`.
+
+````jsonc
+{
+  "fileTypes": [],
+  // This can be something else.
+  "scopeName": "mdx.LANGUAGE.codeblock",
+  "injectionSelector": "L:source.mdx",
+  "patterns": [
+    {
+      // This references the repository key below.
+      "include": "#LANGUAGE-code-block"
+    }
+  ],
+  "repository": {
+    "LANGUAGE-code-block": {
+      "patterns": [
+        {
+          // This adds supports for code blocks using the ```LANGUAGE delimiter.
+          "begin": "(?:^|\\G)[\\t ]*(`{3,})(?:[\\t ]*((?i:(?:.*\\.)?LANGUAGE))(?:[\\t ]+((?:[^\\n\\r`])+))?)(?:[\\t ]*$)",
+          "beginCaptures": {
+            "1": {
+              "name": "string.other.begin.code.fenced.mdx"
+            },
+            "2": {
+              "name": "entity.name.function.mdx"
+            }
+          },
+          "contentName": "meta.embedded.LANGUAGE",
+          "end": "(\\1)(?:[\\t ]*$)",
+          "endCaptures": {
+            "1": {
+              "name": "string.other.end.code.fenced.mdx"
+            }
+          },
+          "name": "markup.code.LANGUAGE.mdx",
+          "patterns": [
+            {
+              "include": "source.LANGUAGE"
+            }
+          ]
+        },
+        {
+          // This adds support for code blocks using the ~~~LANGUAGE delimiter.
+          "begin": "(?:^|\\G)[\\t ]*(~{3,})(?:[\\t ]*((?i:(?:.*\\.)?LANGUAGE))(?:[\\t ]+((?:[^\\n\\r])+))?)(?:[\\t ]*$)",
+          "beginCaptures": {
+            "1": {
+              "name": "string.other.begin.code.fenced.mdx"
+            },
+            "2": {
+              "name": "entity.name.function.mdx"
+            }
+          },
+          "contentName": "meta.embedded.LANGUAGE",
+          "end": "(\\1)(?:[\\t ]*$)",
+          "endCaptures": {
+            "1": {
+              "name": "string.other.end.code.fenced.mdx"
+            }
+          },
+          "name": "markup.code.LANGUAGE.mdx",
+          "patterns": [
+            {
+              "include": "source.LANGUAGE"
```

Comment:
not all languages are `source.*`, perhaps use another variable name, `SCOPE` or so?

Reply:
For Astro it’s even more complex. Perhaps I should write something like:

```jsonc
"patterns": [
  // Your patterns, typically something like:
  // {
  //   "include": "source.LANGUAGE"
  // }
]
```

Reply:
Or, just use a practical example. Such as the mermaid one. And then tell people: change it to the needs of your language.

Reply:
Do you mean to replace `LANGUAGE` with `mermaid` throughout the entire example?

Reply:
yeah something along those lines: say something along the lines of “here’s how Mermaid injects itself into MDX, which you can use as inspiration when doing similar things” and then the code blocks, with actual working code for mermaid

Reply:
Done. IMO it’s neither better nor worse.

The actual Mermaid example implementation slightly different, because they use a build script to include the entire YAML language within the injection. But this exact example would have worked too, it’s what I recommend.